File: DataDistributionManagerNet/SmartDataDistribution.cs
/*
*  Copyright 2021 MASES s.r.l.
*  
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*  
*  http://www.apache.org/licenses/LICENSE-2.0
*  
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
*  
*  Refer to LICENSE for more information.
*/

using MASES.DataDistributionManager.Bindings.Interop;
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace MASES.DataDistributionManager.Bindings
{
    public class ClusterHealthElement
    {
        internal ClusterHealthElement(ClusterHealthElementStruct s)
        {
            this.Status = s.Status;
            this.Uptime = s.Uptime;
            this.LastContactTime = s.LastContactTime;
        }

        public DDM_INSTANCE_STATE Status { get; private set; }
        public Int64 Uptime { get; private set; }
        public Int64 LastContactTime { get; private set; }
    };


    public class SmartDataDistribution : IDataDistributionCallbackLow, IDataDistributionMastershipCallbackLow, ISmartDataDistributionMastership
    {
        public SmartDataDistribution()
        {
            IDataDistribution_ptr = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<DataDistribution_create>().Invoke();
        }

        ~SmartDataDistribution()
        {

        }

        public HRESULT Initialize(string conf_file, string szMyAddress = null, string channelTrailer = null)
        {
            m_DataDistributionCallbackLow = new DataDistributionCallbackLow(IDataDistribution_ptr, this);

            HRESULT result = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistribution_Initialize>().Invoke(
                IDataDistribution_ptr, m_DataDistributionCallbackLow.Pointer,
                conf_file, szMyAddress, channelTrailer);

            if (result.Succeeded) IDataDistributionSubsystemManager_ptr = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistribution_GetSubsystemManager>().Invoke(IDataDistribution_ptr);
            return result;
        }

        public HRESULT Initialize(string[] arrayParams, string szMyAddress = null, string channelTrailer = null)
        {
            m_DataDistributionCallbackLow = new DataDistributionCallbackLow(IDataDistribution_ptr, this);

            HRESULT result = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistribution_Initialize2>().Invoke(
                IDataDistribution_ptr, m_DataDistributionCallbackLow.Pointer,
                arrayParams, arrayParams.Length, szMyAddress, channelTrailer);

            if (result.Succeeded) IDataDistributionSubsystemManager_ptr = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistribution_GetSubsystemManager>().Invoke(IDataDistribution_ptr);
            return result;
        }

        public HRESULT RequestMastershipManager(string serverName, string[] parameters)
        {
            m_DataDistributionMastershipCallbackLow = new DataDistributionMastershipCallbackLow(IDataDistribution_ptr, this);

            return DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistribution_RequestMastershipManager>().Invoke(
                    IDataDistribution_ptr, m_DataDistributionMastershipCallbackLow.Pointer, serverName, parameters, parameters.Length);
        }

        public ISmartDataDistributionMastership MastershipManager { get { return this; } }

        public HRESULT Start(uint dwMilliseconds)
        {
            return DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionSubsystem_Start>().Invoke(IDataDistributionSubsystemManager_ptr, dwMilliseconds);
        }

        public HRESULT Stop(uint dwMilliseconds)
        {
            return DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionSubsystem_Stop>().Invoke(IDataDistributionSubsystemManager_ptr, dwMilliseconds);
        }

        public string GetProtocol()
        {
            IntPtr ptr = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistribution_GetProtocol>().Invoke(IDataDistribution_ptr);
            if (ptr != IntPtr.Zero)
            {
                return Marshal.PtrToStringAnsi(ptr);
            }
            return null;
        }

        public string GetProtocolLib()
        {
            IntPtr ptr = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistribution_GetProtocolLib>().Invoke(IDataDistribution_ptr);
            if (ptr != IntPtr.Zero)
            {
                return Marshal.PtrToStringAnsi(ptr);
            }
            return null;
        }

        public string GetMastershipLib()
        {
            IntPtr ptr = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistribution_GetMastershipLib>().Invoke(IDataDistribution_ptr);
            if (ptr != IntPtr.Zero)
            {
                return Marshal.PtrToStringAnsi(ptr);
            }
            return null;
        }

        public T CreateSmartChannel<T>(string channelName, DDM_CHANNEL_DIRECTION direction = DDM_CHANNEL_DIRECTION.ALL, string[] arrayParams = null)
            where T : SmartDataDistributionChannel
        {
            T smartChannelReference = Activator.CreateInstance(typeof(T)) as T;

            IntPtr handle = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionSubsystem_CreateChannel>().Invoke(
                IDataDistributionSubsystemManager_ptr, channelName, smartChannelReference.m_DataDistributionChannelCallbackLow.Pointer, direction, arrayParams, (arrayParams != null) ? arrayParams.Length : 0);

            smartChannelReference.m_direction = direction;
            smartChannelReference.IDataDistributionSubsystemManager_ptr = IDataDistributionSubsystemManager_ptr;
            smartChannelReference.channelHandle = handle;

            return smartChannelReference;
        }

        #region IDataDistributionCallback
        string IDataDistributionCallbackLow.OnConfiguration(IntPtr IDataDistribution_nativePtr, string key, string value)
        {
            return this.OnConfiguration(IDataDistribution_nativePtr, key, value);
        }

        void IDataDistributionCallbackLow.OnLogging(IntPtr IDataDistribution_nativePtr, DDM_LOG_LEVEL level, string source, string function, string errStr)
        {
            this.OnLogging(IDataDistribution_nativePtr, level, source, function, errStr);
        }

        void IDataDistributionCallbackLow.OnCompletelyDisconnected(IntPtr IDataDistribution_nativePtr, string channelName, string reason)
        {
            this.OnCompletelyDisconnected(IDataDistribution_nativePtr, channelName, reason);
        }

        public virtual string OnConfiguration(IntPtr IDataDistribution_nativePtr, string key, string value)
        {
            return value;
        }

        public virtual void OnLogging(IntPtr IDataDistribution_nativePtr, DDM_LOG_LEVEL level, string source, string function, string errStr)
        {

        }

        public virtual void OnCompletelyDisconnected(IntPtr IDataDistribution_nativePtr, string channelName, string reason)
        {

        }
        #endregion

        #region IDataDistributionMastershipCallback
        public virtual void OnClusterStateChange(IntPtr IDataDistribution_nativePtr, DDM_CLUSTEREVENT change, Int64 serverid)
        {

        }

        public virtual void OnStateChange(IntPtr IDataDistribution_nativePtr, DDM_INSTANCE_STATE newState, DDM_INSTANCE_STATE oldState)
        {

        }

        public virtual void OnStateReady(IntPtr IDataDistribution_nativePtr, IntPtr buffer, Int64 len)
        {

        }

        public virtual void OnRequestedState(IntPtr IDataDistribution_nativePtr, IntPtr buffer, IntPtr len)
        {

        }

        public virtual void OnMultiplePrimary(IntPtr IDataDistribution_nativePtr, Int64 myId, Int64 otherId)
        {

        }

        public virtual void FirstStateChange(IntPtr IDataDistribution_nativePtr, DDM_INSTANCE_STATE newState)
        {

        }

        public virtual void ChangingState(IntPtr IDataDistribution_nativePtr, DDM_INSTANCE_STATE oldState, DDM_INSTANCE_STATE newState)
        {

        }

        public virtual void ChangedState(IntPtr IDataDistribution_nativePtr, DDM_INSTANCE_STATE newState)
        {

        }

        #endregion

        #region ISmartDataDistributionMastership

        HRESULT ISmartDataDistributionMastership.Start(uint dwMilliseconds)
        {
            return DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_Start>().Invoke(IDataDistribution_ptr, dwMilliseconds);
        }

        HRESULT ISmartDataDistributionMastership.Stop(uint dwMilliseconds)
        {
            return DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_Stop>().Invoke(IDataDistribution_ptr, dwMilliseconds);
        }

        void ISmartDataDistributionMastership.ChangeMyState(DDM_INSTANCE_STATE newState)
        {
            DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_ChangeMyState>().Invoke(IDataDistribution_ptr, newState);
        }

        void ISmartDataDistributionMastership.ChangeState(Int64 instanceId, DDM_INSTANCE_STATE newState)
        {
            DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_ChangeState>().Invoke(IDataDistribution_ptr, instanceId, newState);
        }

        bool ISmartDataDistributionMastership.RequestIAmNextPrimary()
        {
            return DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_RequestIAmNextPrimary>().Invoke(IDataDistribution_ptr);
        }

        bool ISmartDataDistributionMastership.IamNextPrimary
        {
            get
            {
                return DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_GetIamNextPrimary>().Invoke(IDataDistribution_ptr);
            }
        }

        IDictionary<Int64, DDM_INSTANCE_STATE> ISmartDataDistributionMastership.ClusterState
        {
            get
            {
                IDictionary<Int64, DDM_INSTANCE_STATE> result = new Dictionary<Int64, DDM_INSTANCE_STATE>();

                Int64[] arrayElements = null;
                IntPtr length = IntPtr.Zero;

                DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_GetClusterIndexes>().Invoke(IDataDistribution_ptr, arrayElements, length);

                if (length != IntPtr.Zero && arrayElements != null)
                {
                    foreach (var item in arrayElements)
                    {
                        var res = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_GetStateOf>().Invoke(IDataDistribution_ptr, item);
                        result.Add(item, res);
                    }
                }

                return result;
            }
        }

        IDictionary<Int64, ClusterHealthElement> ISmartDataDistributionMastership.ClusterHealth
        {
            get
            {
                IDictionary<Int64, ClusterHealthElement> result = new Dictionary<Int64, ClusterHealthElement>();

                Int64[] arrayElements = null;
                IntPtr length = IntPtr.Zero;

                DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_GetClusterIndexes>().Invoke(IDataDistribution_ptr, arrayElements, length);

                if (length != IntPtr.Zero && arrayElements != null)
                {
                    foreach (var item in arrayElements)
                    {
                        var res = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_GetClusterHealth>().Invoke(IDataDistribution_ptr, item);
                        var structure = Marshal.PtrToStructure<ClusterHealthElementStruct>(res);
                        result.Add(item, new ClusterHealthElement(structure));
                    }
                }

                return result;
            }
        }

        DDM_INSTANCE_STATE ISmartDataDistributionMastership.MyState
        {
            get
            {
                return DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_GetMyState>().Invoke(IDataDistribution_ptr);
            }
        }

        Int64 ISmartDataDistributionMastership.LocalServerId
        {
            get
            {
                Int64 res = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_GetLocalServerId>().Invoke(IDataDistribution_ptr);
                return res;
            }
        }

        Int64 ISmartDataDistributionMastership.PrimaryServerId
        {
            get
            {
                Int64 res = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_GetPrimaryServerId>().Invoke(IDataDistribution_ptr);
                return res;
            }
        }

        Int64 ISmartDataDistributionMastership.MessageDelay
        {
            get
            {
                Int64 res = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_GetMessageDelay>().Invoke(IDataDistribution_ptr);
                return res;
            }
        }

        Int64 ISmartDataDistributionMastership.UpTime
        {
            get
            {
                Int64 res = DataDistributionManagerInvokeWrapper.DataDistributionEnv.GetDelegate<IDataDistributionMastershipCommon_GetUpTime>().Invoke(IDataDistribution_ptr);
                return res;
            }
        }

        #endregion

        DataDistributionCallbackLow m_DataDistributionCallbackLow;
        DataDistributionMastershipCallbackLow m_DataDistributionMastershipCallbackLow;
        IntPtr IDataDistribution_ptr;
        IntPtr IDataDistributionSubsystemManager_ptr;
    }
}

File: DataDistributionManagerJava/src/org/mases/datadistributionmanager/SmartDataDistribution.java
/*
*  Copyright 2021 MASES s.r.l.
*
*  Licensed under the Apache License, Version 2.0 (the "License");
*  you may not use this file except in compliance with the License.
*  You may obtain a copy of the License at
*
*  http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing, software
*  distributed under the License is distributed on an "AS IS" BASIS,
*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*  See the License for the specific language governing permissions and
*  limitations under the License.
*
*  Refer to LICENSE for more information.
*/

package org.mases.datadistributionmanager;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.*;
import java.io.BufferedOutputStream;
import java.io.FileInputStream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class SmartDataDistribution implements IDataDistributionCallbackLow, IDataDistributionMastershipCallbackLow {
    long m_DataDistributionCallbackLow;
    long m_DataDistributionMastershipCallbackLow;
    long IDataDistribution_ptr;
    long IDataDistributionSubsystemManager_ptr;
    long m_InitializeResult = -1;

    private static void unzip(String zipFilePath, String destDirectory) throws IOException {
        File destDir = new File(destDirectory);
        if (!destDir.exists()) {
            destDir.mkdir();
        }
        try (ZipInputStream zipIn = new ZipInputStream(new FileInputStream(zipFilePath))) {
            ZipEntry entry = zipIn.getNextEntry();
            while (entry != null) {
                String filePath = destDirectory + File.separator + entry.getName();
                if (!entry.isDirectory()) {
                    extractFile(zipIn, filePath);
                    if (!System.getProperty("os.name").contains("Windows")) {
                        try {
                            Runtime.getRuntime().exec(new String[] { "chmod", "755", filePath }).waitFor();
                        } catch (Throwable e) {
                        }
                    }
                } else {
                    File dir = new File(filePath);
                    dir.mkdirs();
                }
                zipIn.closeEntry();
                entry = zipIn.getNextEntry();
            }
        }
    }

    private static void extractFile(ZipInputStream zipIn, String filePath) throws IOException {
        final int BUFFER_SIZE = 4096;
        try (BufferedOutputStream bos = new BufferedOutputStream(new FileOutputStream(filePath))) {
            byte[] bytesIn = new byte[BUFFER_SIZE];
            int read = 0;
            while ((read = zipIn.read(bytesIn)) != -1) {
                bos.write(bytesIn, 0, read);
            }
        }
    }

    private static boolean extractAndLoadLibraryFile(String libFolderForCurrentOS, String targetFolder,
            String entryLibrary) {
        final String containerFileName = "nativepackage.zip";

        String nativeLibraryFilePath = libFolderForCurrentOS + "/" + containerFileName;

        String extractedLibFileName = containerFileName;
        File extractedLibFile = new File(targetFolder, extractedLibFileName);

        try {
            // Extract file into the current directory
            InputStream reader = SmartDataDistribution.class.getResourceAsStream(nativeLibraryFilePath);
            FileOutputStream writer = new FileOutputStream(extractedLibFile);
            byte[] buffer = new byte[1024];
            int bytesRead = 0;
            while ((bytesRead = reader.read(buffer)) != -1) {
                writer.write(buffer, 0, bytesRead);
            }

            writer.close();
            reader.close();

            unzip(extractedLibFile.toString(), targetFolder);

            return loadNativeLibrary(targetFolder, entryLibrary);
        } catch (IOException e) {
            System.err.println(e.getMessage());
            return false;
        }
    }

    private static synchronized boolean loadNativeLibrary(String path, String name) {
        File libPath = new File(path, name);
        if (libPath.exists()) {
            try {
                System.load(new File(path, name).getAbsolutePath());
                return true;
            } catch (UnsatisfiedLinkError e) {
                System.err.println(e);
                return false;
            }

        } else
            return false;
    }

    static boolean LoadWrapper() {
        String osName = System.getProperty("os.name");
        String osArch = System.getProperty("os.arch");

        String libFolderForCurrentOS = "windows_x64";
        String libName = "DataDistributionManager.dll";

        if (osName.toLowerCase().contains("linux")) {
            if (osArch.toLowerCase().contains("amd64") || osArch.toLowerCase().contains("x86_64")) {
                libFolderForCurrentOS = "linux_x64";
                libName = "DataDistributionManager.so";
            } else if (osArch.toLowerCase().contains("x86")) {
                libFolderForCurrentOS = "linux_x86";
                libName = "DataDistributionManager.so";
            } else if (osArch.toLowerCase().contains("arm")) {
                libFolderForCurrentOS = "linux_arm";
                libName = "DataDistributionManager.so";
            } else if (osArch.toLowerCase().contains("arm64")) {
                libFolderForCurrentOS = "linux_arm64";
                libName = "DataDistributionManager.so";
            }
        } else if (osName.toLowerCase().contains("mac")) {
            if (osArch.toLowerCase().contains("64")) {
                libFolderForCurrentOS = "mac_x64";
                libName = "DataDistributionManager.dylib";
            }
        } else if (osName.toLowerCase().contains("windows")) {
            if (osArch.toLowerCase().contains("amd64") || osArch.toLowerCase().contains("x86_64")) {
                libFolderForCurrentOS = "windows_x64";
                libName = "DataDistributionManager.dll";
            } else if (osArch.toLowerCase().contains("x86")) {
                libFolderForCurrentOS = "windows_x86";
                libName = "DataDistributionManager.dll";
            } else if (osArch.toLowerCase().contains("arm")) {
                libFolderForCurrentOS = "windows_arm";
                libName = "DataDistributionManager.dll";
            } else if (osArch.toLowerCase().contains("arm64")) {
                libFolderForCurrentOS = "windows_arm64";
                libName = "DataDistributionManager.dll";
            }
        }

        try {
            Path tmpPath = Files.createTempDirectory(SmartDataDistribution.class.getName());
            return extractAndLoadLibraryFile(libFolderForCurrentOS, tmpPath.toString(), libName);
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
    }

    static final boolean loaded = LoadWrapper();

    public SmartDataDistribution() {
        IDataDistribution_ptr = NativeInterface.DataDistribution_create();
    }

    public HRESULT Initialize(String conf_file) throws IllegalArgumentException {
        return Initialize(conf_file, null, null);
    }

    public HRESULT Initialize(String conf_file, String topicTrailer) throws IllegalArgumentException {
        return Initialize(conf_file, null, topicTrailer);
    }

    public HRESULT Initialize(String conf_file, String szMyAddress, String topicTrailer) {
        m_DataDistributionCallbackLow = NativeCallbackManager.RegisterCallback((IDataDistributionCallbackLow) this);

        m_InitializeResult = NativeInterface.IDataDistribution_Initialize(IDataDistribution_ptr,
                m_DataDistributionCallbackLow, conf_file, szMyAddress, topicTrailer);

        if (m_InitializeResult == 0) {
            IDataDistributionSubsystemManager_ptr = NativeInterface
                    .IDataDistribution_GetSubsystemManager(IDataDistribution_ptr);
        } else {
            NativeCallbackManager.UnregisterCallback((IDataDistributionCallbackLow) this);
        }
        return new HRESULT(m_InitializeResult);
    }

    public HRESULT Initialize(IConfiguration configuration) throws IllegalArgumentException {
        return Initialize(configuration.getConfiguration(), null, null);
    }

    public HRESULT Initialize(IConfiguration configuration, String topicTrailer) throws IllegalArgumentException {
        return Initialize(configuration.getConfiguration(), null, topicTrailer);
    }

    public HRESULT Initialize(IConfiguration configuration, String szMyAddress, String topicTrailer)
            throws IllegalArgumentException {
        return Initialize(configuration.getConfiguration(), szMyAddress, topicTrailer);
    }

    public HRESULT Initialize(String[] arrayParams) throws IllegalArgumentException {
        return Initialize(arrayParams, null, null);
    }

    public HRESULT Initialize(String[] arrayParams, String topicTrailer) throws IllegalArgumentException {
        return Initialize(arrayParams, null, topicTrailer);
    }

    public HRESULT Initialize(String[] arrayParams, String szMyAddress, String topicTrailer) {
        m_DataDistributionCallbackLow = NativeCallbackManager.RegisterCallback((IDataDistributionCallbackLow) this);

        m_InitializeResult = NativeInterface.IDataDistribution_Initialize2(IDataDistribution_ptr,
                m_DataDistributionCallbackLow, arrayParams, szMyAddress, topicTrailer);

        if (m_InitializeResult == 0) {
            IDataDistributionSubsystemManager_ptr = NativeInterface
                    .IDataDistribution_GetSubsystemManager(IDataDistribution_ptr);
        } else {
            NativeCallbackManager.UnregisterCallback((IDataDistributionCallbackLow) this);
        }
        return new HRESULT(m_InitializeResult);
    }

    public HRESULT RequestMastershipManager(String serverName, String[] parameters) {
        if (m_SmartDataDistributionMastership != null)
            return HRESULT.S_OK;

        m_DataDistributionMastershipCallbackLow = NativeCallbackManager
                .RegisterCallback((IDataDistributionMastershipCallbackLow) this);

        long res = NativeInterface.IDataDistribution_RequestMastershipManager(IDataDistribution_ptr,
                m_DataDistributionMastershipCallbackLow, serverName, parameters);

        HRESULT hRes = new HRESULT(res);

        if (hRes == HRESULT.S_OK) {
            m_SmartDataDistributionMastership = new SmartDataDistributionMastership(IDataDistribution_ptr);
        } else {
            NativeCallbackManager.UnregisterCallback((IDataDistributionMastershipCallbackLow) this);
        }

        return hRes;
    }

    SmartDataDistributionMastership m_SmartDataDistributionMastership;

    public ISmartDataDistributionMastership getMastershipManager() {
        return m_SmartDataDistributionMastership;
    }

    public HRESULT Start(int dwMilliseconds) {
        long res = -1;
        if (IDataDistributionSubsystemManager_ptr != 0) {
            res = NativeInterface.IDataDistributionSubsystem_Start(IDataDistributionSubsystemManager_ptr,
                    dwMilliseconds);
        }
        return new HRESULT(res);
    }

    public HRESULT Stop(int dwMilliseconds) {
        long res = NativeInterface.IDataDistributionSubsystem_Stop(IDataDistributionSubsystemManager_ptr,
                dwMilliseconds);
        return new HRESULT(res);
    }

    public String GetProtocol() {
        if (IDataDistributionSubsystemManager_ptr != 0) {
            return NativeInterface.IDataDistribution_GetProtocol(IDataDistributionSubsystemManager_ptr);
        }
        return "";
    }

    public String GetProtocolLib() {
        if (IDataDistributionSubsystemManager_ptr != 0) {
            return NativeInterface.IDataDistribution_GetProtocolLib(IDataDistributionSubsystemManager_ptr);
        }
        return "";
    }

    public String GetMastershipLib() {
        if (IDataDistributionSubsystemManager_ptr != 0) {
            return NativeInterface.IDataDistribution_GetMastershipLib(IDataDistributionSubsystemManager_ptr);
        }
        return "";
    }

    public <T extends SmartDataDistributionChannel> T CeateSmartChannel(Class<T> clazz, String topicName)
            throws Throwable {
        return CeateSmartChannel(clazz, topicName, DDM_CHANNEL_DIRECTION.ALL, (IConfiguration) null);
    }

    public <T extends SmartDataDistributionChannel> T CeateSmartChannel(Class<T> clazz, String topicName,
            DDM_CHANNEL_DIRECTION direction) throws Throwable {
        return CeateSmartChannel(clazz, topicName, direction, (IConfiguration) null);
    }

    public <T extends SmartDataDistributionChannel> T CeateSmartChannel(Class<T> clazz, String topicName,
            DDM_CHANNEL_DIRECTION direction, IConfiguration configuration) throws Throwable {
        return CeateSmartChannel(clazz, topicName, direction,
                (configuration == null) ? null : configuration.getConfiguration());
    }

    @SuppressWarnings("unchecked")
    public <T extends SmartDataDistributionChannel> T CeateSmartChannel(Class<T> clazz, String topicName,
            DDM_CHANNEL_DIRECTION direction, String[] arrayParams) throws Throwable {
        if (IDataDistributionSubsystemManager_ptr == 0)
            return null;

        Object newObject = Class.forName(clazz.getName()).getDeclaredConstructor().newInstance();
        SmartDataDistributionChannel inm = (SmartDataDistributionChannel) newObject;

        long handle = NativeInterface.IDataDistributionSubsystem_CreateChannel(IDataDistributionSubsystemManager_ptr,
                topicName, inm.m_DataDistributionChannelCallbackLow, direction.atomicNumber, arrayParams);

        if (handle == 0) {
            throw new Exception("Unable to create channel, see log for the reason.");
        }

        inm.m_direction = direction;
        inm.IDataDistributionSubsystemManager_ptr = IDataDistributionSubsystemManager_ptr;
        inm.topicHandle = handle;

        return (T) inm;
    }

    public final String OnConfiguration(long IDataDistribution_nativePtr, String key, String value) {
        return this.OnConfiguration(key, value);
    }

    public String OnConfiguration(String key, String value) {
        return value;
    }

    public final void OnLogging(long IDataDistribution_nativePtr, int level, String source, String function,
            String errStr) {
        this.OnLogging(DDM_LOG_LEVEL.valueOfAtomicNumber(level), source, function, errStr);
    }

    public void OnLogging(DDM_LOG_LEVEL level, String source, String function, String errStr) {

    }

    public void OnCompletelyDisconnected(long IDataDistribution_nativePtr, String topicName, String reason) {
        OnCompletelyDisconnected(topicName, reason);
    }

    public void OnCompletelyDisconnected(String topicName, String reason) {

    }

    // IDataDistributionMastershipCallback
    public final void OnClusterStateChange(long IDataDistribution_nativePtr, int change, long serverid) {
        OnClusterStateChange(DDM_CLUSTEREVENT.valueOfAtomicNumber(change), serverid);
    }

    public void OnClusterStateChange(DDM_CLUSTEREVENT change, long serverid) {

    }

    public final void OnStateChange(long IDataDistribution_nativePtr, int newState, int oldState) {
        OnStateChange(DDM_INSTANCE_STATE.valueOfAtomicNumber(newState),
                DDM_INSTANCE_STATE.valueOfAtomicNumber(oldState));
    }

    public void OnStateChange(DDM_INSTANCE_STATE newState, DDM_INSTANCE_STATE oldState) {

    }

    public final void OnStateReady(long IDataDistribution_nativePtr, byte[] buffer) {
        OnStateReady(buffer);
    }

    public void OnStateReady(byte[] buffer) {

    }

    public final void OnRequestedState(long IDataDistribution_nativePtr, byte[] buffer) {
        OnRequestedState(buffer);
    }

    public void OnRequestedState(byte[] buffer) {

    }

    public final void OnMultiplePrimary(long IDataDistribution_nativePtr, long myId, long otherId) {
        OnMultiplePrimary(myId, otherId);
    }

    public void OnMultiplePrimary(long myId, long otherId) {

    }

    public final void FirstStateChange(long IDataDistribution_nativePtr, int newState) {
        FirstStateChange(DDM_INSTANCE_STATE.valueOfAtomicNumber(newState));
    }

    public void FirstStateChange(DDM_INSTANCE_STATE newState) {

    }

    public final void ChangingState(long IDataDistribution_nativePtr, int oldState, int newState) {
        ChangingState(DDM_INSTANCE_STATE.valueOfAtomicNumber(oldState),
                DDM_INSTANCE_STATE.valueOfAtomicNumber(newState));
    }

    public void ChangingState(DDM_INSTANCE_STATE oldState, DDM_INSTANCE_STATE newState) {

    }

    public final void ChangedState(long IDataDistribution_nativePtr, int newState) {
        ChangedState(DDM_INSTANCE_STATE.valueOfAtomicNumber(newState));
    }

    public void ChangedState(DDM_INSTANCE_STATE newState) {

    }
}
